lib/charms/vault_k8s/v0/juju_facade.py

```diff
@@ -436,57 +476,106 @@ def get_remote_app_relation_data(
+        self,
+        relation_name: str,
+        relation_id: int | None = None,
+        relation: Relation | None = None,
     ) -> List[RelationDataContent | dict[str, str]]:
         """Get relation data from the remote units databags.
 
         Raises:
             NoSuchRelationError
             MultipleRelationsFoundError
+            ValueError
         """
+        if not relation and not relation_id:
+            raise ValueError("Either relation or relation_id must be provided")
         relation = (
-            self.get_relation_by_id(relation_name, relation_id)
-            if relation_id
-            else self.get_relation_by_name(relation_name)
+            (
+                self.get_relation_by_id(relation_name, relation_id)
+                if relation_id
+                else self.get_relation_by_name(relation_name)
+            )
+            if not relation
+            else relation
         )
         return [relation.data.get(unit, {}) for unit in relation.units]
 
     def _set_relation_data(
         self,
         data: dict[str, str],
         relation_name: str,
-        relation_id: int | None,
         entity: Unit | Application,
+        relation_id: int | None = None,
+        relation: Relation | None = None,
     ) -> None:
+        if not relation and not relation_id:
+            raise ValueError("Either relation or relation_id must be provided")
         relation = (
-            self.get_relation_by_id(relation_name, relation_id)
-            if relation_id
-            else self.get_relation_by_name(relation_name)
+            (
+                self.get_relation_by_id(relation_name, relation_id)
+                if relation_id
+                else self.get_relation_by_name(relation_name)
+            )
+            if not relation
+            else relation
         )
         if not all(isinstance(value, str) for value in chain(data.values(), data.keys())):
             raise InvalidRelationDataError("Invalid relation data")
         try:
-            logger.info("Setting relation data for %s:%d", relation_name, relation_id)
+            logger.info(
+                "Setting relation data for %s:%d", relation_name, relation_id or relation.id
+            )
             relation.data[entity].update(data)
         except RelationDataError as e:
             logger.error(
-                "Error setting relation data for %s:%d: %s", relation_name, relation_id, e
+                "Error setting relation data for %s:%d: %s",
+                relation_name,
+                relation_id or relation.id,
+                e,
             )
             raise InvalidRelationDataError(e) from e
 
     def set_app_relation_data(
-        self, data: dict[str, str], relation_name: str, relation_id: int | None
+        self,
+        data: dict[str, str],
+        relation_name: str,
+        relation_id: int | None = None,
+        relation: Relation | None = None,
```

Comment:
The `relation` prefixes are unnecessary, it's obvious we are talking about the relation.

Reply:
Removed